### PR TITLE
Add VM snapshot modal and delete resource modal

### PR DESCRIPTION
**Pull request by avivtur**
## 📝 Description

This PR adds: 
Create VM snapshot modal (including deadline as missing from 4.10)
Adding DeleteResourceModal, which serves the purpose of delete resource confirmation modal.
Edit: 
Adding Restore snapshot button and modal

## 🎥 Demo

**UPDATED***

https://user-images.githubusercontent.com/67270715/158056616-7d18e0cc-6941-4dcd-95f2-cb2e0fc34641.mp4

@glekner @metalice @vojtechszocs @yaacov @pcbailey @upalatucci @hstastna please review

### CODE REVIEW

File: src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx [hunk 1]
```diff
@@ -18,17 +23,38 @@ type SnapshotListPageProps = RouteComponentProps<{
...
-        <SnapshotList vm={obj} />
+        <SnapshotList
+          snapshots={snapshots}
+          restoresMap={restoresMap}
+          loaded={loaded}
+          error={error}
+          isVMRunning={vm?.status?.printableStatus !== printableVMStatus.Stopped}
+        />
       </ListPageBody>
+      {isOpen && (
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx:50]: why do we need this?

File: src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx [hunk 2]
```diff
@@ -18,17 +23,38 @@ type SnapshotListPageProps = RouteComponentProps<{
...
+        <SnapshotModal
+          usedNames={getUsedSnapshotNames(snapshots)}
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx:52]: why do we need this client side logic? the backend should reject a request if a snapshot with the same name is being created

File: src/views/virtualmachines/details/tabs/snapshots/components/list/SnapshotRow.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        <Tooltip content={t('Restore is enabled only for offline virtual machine.')}>
+          <Button
+            isAriaDisabled={isRestoreDisabled}
+            variant="secondary"
+            onClick={() => setIsOpen(true)}
+          >
+            {t('Restore')}
+          </Button>
+        </Tooltip>
+        {isOpen && (
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/components/list/SnapshotRow.tsx:61]: again. why do we need this line?

File: src/views/virtualmachines/details/tabs/snapshots/components/modal/SnapshotFormFields/SnapshotNameFormField.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  usedNames,
+  setIsError,
+}) => {
+  const { t } = useKubevirtTranslation();
+
+  const [snapshotNameError, setSnapshotNameError] = React.useState(undefined);
+
+  const handleNameChange = (value: string, event: React.FormEvent<HTMLInputElement>) => {
+    event.preventDefault();
+    const error = validateSnapshotName(value, usedNames, t);
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/components/modal/SnapshotFormFields/SnapshotNameFormField.tsx:28]: we don't need this, the request should fail from the backend

File: src/views/virtualmachines/details/tabs/snapshots/utils/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+export const createSnapshotName = () => {
+  return `snapshot-${uniqueNamesGenerator({
+    dictionaries: [adjectives, animals],
+    separator: '-',
+  })}`;
+};
+
+export const generateSnapshotName = (usedNames: string[]): string => {
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/utils/helpers.ts:32]: no need for `usedNames`

File: src/views/virtualmachines/details/tabs/snapshots/utils/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  let generatedName = createSnapshotName();
+  while (usedNames?.includes(generatedName)) {
+    generatedName = createSnapshotName();
+  }
+  return generatedName;
+};
+
+export const validateSnapshotName = (name: string, usedNames: string[], t: TFunction): string => {
```
Comment (glekner) [inline src/views/virtualmachines/details/tabs/snapshots/utils/helpers.ts:40]: we don't need this

**Review by glekner — COMMENTED** [review]
nice job, left some comments

File: src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx (hunk 1 above)
Comment (avivtur) [reply src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx:50]: to get a new render cycle for the modal component so the data in the form state will be renewed on closing the modal,
any thoughts on how to improve this?

File: src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx (hunk 2 above)
Comment (avivtur) [reply src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx:52]: np, will remove

File: src/views/virtualmachines/details/tabs/snapshots/components/list/SnapshotRow.tsx (hunk 3 above)
Comment (avivtur) [reply src/views/virtualmachines/details/tabs/snapshots/components/list/SnapshotRow.tsx:61]: replied above

File: src/views/virtualmachines/details/tabs/snapshots/components/modal/SnapshotFormFields/SnapshotNameFormField.tsx (hunk 4 above)
Comment (avivtur) [reply src/views/virtualmachines/details/tabs/snapshots/components/modal/SnapshotFormFields/SnapshotNameFormField.tsx:28]: so no need for any validation? how about some trivial checks like that the input is not emtpy?

Comment (glekner) [reply src/views/virtualmachines/details/tabs/snapshots/components/modal/SnapshotFormFields/SnapshotNameFormField.tsx:28]: no need, we don't want client side validation. only if its really required

File: src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx (hunk 1 above)
Comment (glekner) [reply src/views/virtualmachines/details/tabs/snapshots/SnapshotListPage.tsx:50]:
```tsx
  // reset state when modal is closed
  React.useEffect(() => {
    setState(initialState);
  }, [isOpen]);
```

**Pull request merged**